src/store/subscription_impl.rs

```diff
@@ -0,0 +1,409 @@
+//! The users should create a SubscriptionImpl instance and pass it to the store as hooks and give it
+//! to the downstream services to subscribe to the invoice/payment updates.
+
+use std::collections::{hash_map::Entry, HashMap};
+
+use ractor::{
+    async_trait, call_t, Actor, ActorProcessingErr, ActorRef, DerivedActorRef, MessagingErr,
+    RactorErr, RpcReplyPort,
+};
+use thiserror::Error;
+use tracing::warn;
+
+use crate::{
+    fiber::{
+        graph::{NetworkGraphStateStore, PaymentSessionStatus},
+        types::Hash256,
+    },
+    invoice::{CkbInvoiceStatus, InvoiceStore},
+};
+
+use super::{
+    hook::{InvoiceUpdateHook, PaymentUpdateHook},
+    subscription::{
+        InvoiceState, InvoiceSubscription, InvoiceUpdate, PaymentState, PaymentSubscription,
+        PaymentUpdate, StoreUpdateSubscription,
+    },
+    Store,
+};
+
+const CALLING_ACTOR_TIMEOUT_MS: u64 = 1000;
+
+pub(crate) struct SubscriptionActor {
+    store: Store,
+}
+
+impl SubscriptionActor {
+    pub fn new(store: Store) -> Self {
+        Self { store }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct SubscriptionImpl {
+    actor: ActorRef<SubscriptionActorMessage>,
+}
+
+pub async fn new_subscription_impl(store: Store) -> SubscriptionImpl {
+    let actor = SubscriptionActor::new(store);
+    let actor_ref = Actor::spawn(None, actor, ())
+        .await
+        .expect("start store subscription actor")
+        .0;
+    SubscriptionImpl { actor: actor_ref }
+}
+
+#[derive(Debug)]
+pub enum SubscriptionActorMessage {
+    InvoiceUpdated(Hash256, CkbInvoiceStatus),
+    PaymentUpdated(Hash256, PaymentSessionStatus),
+    SubscribeInvoiceUpdates(
+        Hash256,
+        DerivedActorRef<InvoiceUpdate>,
+        RpcReplyPort<SubscriptionId>,
+    ),
+    UnsubscribeInvoiceUpdates(SubscriptionId),
+    SubscribePaymentUpdates(
+        Hash256,
+        DerivedActorRef<PaymentUpdate>,
+        RpcReplyPort<SubscriptionId>,
+    ),
+    UnsubscribePaymentUpdates(SubscriptionId),
+}
+
+pub type SubscriptionId = u64;
+
+struct InvoiceSubscriber {
+    id: SubscriptionId,
+    receiver: DerivedActorRef<InvoiceUpdate>,
+}
+
+impl InvoiceSubscriber {
+    pub fn new(id: SubscriptionId, receiver: DerivedActorRef<InvoiceUpdate>) -> Self {
+        Self { id, receiver }
+    }
+
+    pub fn send_update(&self, update: InvoiceUpdate) -> bool {
+        self.receiver.send_message(update).is_ok()
+    }
+}
+
+struct PaymentSubscriber {
+    id: SubscriptionId,
+    receiver: DerivedActorRef<PaymentUpdate>,
+}
+
+impl PaymentSubscriber {
+    pub fn new(id: SubscriptionId, receiver: DerivedActorRef<PaymentUpdate>) -> Self {
+        Self { id, receiver }
+    }
+
+    pub fn send_update(&self, update: PaymentUpdate) -> bool {
+        self.receiver.send_message(update).is_ok()
+    }
+}
+
+#[derive(Default)]
+pub struct SubscriptionActorState {
+    next_subscriber_id: SubscriptionId,
+    invoice_subscriptions: HashMap<Hash256, Vec<InvoiceSubscriber>>,
```

Comment:
I think a HashMap is simpler:
```suggestion
    invoice_subscriptions: HashMap<Hash256, HashMap<SubscriptionId, DerivedActorRef<InvoiceUpdate>>>,
```